docs/specs/src/hwp_4_formal.md

```diff
@@ -0,0 +1,403 @@
+
+There are three main types of Zome functions:
+1. ( $z _f$ ) zome functions which do not alter state.
+2. ( $Z _f$ ) that can be called to produce state changes, as well as the 
+3. Validation Rules ( $V _R$ ) for enforcing data integrity of any such state changes (additions, modifications, or deletions of data). 
+
+$$
+\begin{aligned}
+ z _{f_1} \dots z _{f_x} &\in \text{Coordinator DNA} \\
+ Z _{f_1} \dots Z _{f_x} &\in \text{Coordinator DNA} \\
+ V _{R_1} \dots V _{R_x} &\in \text{Integrity DNA} 
+\end{aligned}
+$$
+
+*Note about Functions: Most zome calls don't need to be in the immutable mutually enforced rules included in the DNA hash (Integrity Zomes) because since there's no way to verify what functions were involved in producing data, other nodes can only look at the outputs produced to ensure data is structured and validated correctly.*
+
+The first entry of each agent's source chain contains the app DNA. The hash of this initial DNA entry is what demonstrates that each agent possessed identical and complete copies of the the rules by which they will manage and mutually enforce all state changes. If a node has a different DNA hash, then it is in a different network playing by a different set of rules.
+
+**Genesis:** The genesis process for each agent creates three initial entries. 
+
+1. The DNA is stored in the first chain entry with action $C _0$ like this:
+
+$$
+C _0 = WASM
+\begin{Bmatrix}
+h {_{DNA}} \\
+e _\text{DNA}
+\end{Bmatrix}
+$$
+
+2. Followed by a "Proof of Membership" which other nodes can use to validate whether the agent is allowed to join the application network. It can be left empty if the application membrane is completely open and it doesn't check or use proofs of membership.
+
+$$
+C _1 = 
+\begin{Bmatrix}
+    h _\text{jp} \\ 
+    e _\text{jp}
+\end{Bmatrix}
+$$
+
+3. And finally the agent's Public Key that they have generated, which also becomes their address on the network and DHT. Keys are the only entry type for which the hash algorithm is equality (meaning the hash of a key is the key itself, so it cannot contain any content other than the public key).
+
+$$
+C _2 = 
+\begin{Bmatrix}
+    h _\text{K}  \\ 
+    e _\text{K}
+\end{Bmatrix}
+$$
+
+**Initialization:** After genesis, DNAs may have also provided initialization functions which are all executed the first time an inbound zome call is received and run. This delay in initialization is to allow time for the application to have joined and been validated into the network, just in case initialization functions may need to retrieve some data settings from the network.
+
+Initialization functions may write entries to the chain, or send messages or any variety of actions, but after each DNA and each zome (according the order they were bundled together) have succesfully completed their initializations, an Init_Complete entry is written to the source chain, so that it will not re-attempt initialization to prevent any redundant side-effects.
```

Comment:
```suggestion
**Initialization:** After genesis, DNAs may have also provided initialization functions which are all executed the first time an inbound zome call is received and run. Initialization functions may write entries to the chain, or send messages or any variety of actions, but after each DNA and each zome (according the order they were bundled together) have succesfully completed their initializations, an Init_Complete entry is written to the source chain, so that it will not re-attempt initialization to prevent any redundant side-effects.
```

In the formal spec, let's leave out unnecessary implementation details that may change. We have discussed the possibility of running initialization automatically some day. The current reality of running during the first zome call is a less-than-ideal convenience that was easy to implement, so let's not hold ourselves to it.